src/wms/src/main/java/org/geoserver/wms/featureinfo/FeatureInfoStylePreprocessor.java

```diff
@@ -242,6 +243,9 @@ public void visit(Rule rule) {
         geometriesOnTextSymbolizer.removeAll(geometriesOnLineSymbolizer);
         geometriesOnTextSymbolizer.removeAll(geometriesOnPointSymbolizer);
         for (Expression geom : geometriesOnTextSymbolizer) {
+            if (Function.class.isAssignableFrom(geom.getClass())) {
```

Comment:
Lacks a test. Also, not sure it's doing what it's meant to be doing. The objective of this code is to allow people hitting geometries that are just labelled, but not painted in any other way.
If the label point is used, then we should create a symbolizer for it anyways. 

In other words, the fix is avoiding the exception, but still now allowing GetFeatureInfo to work when the gemetry is just labelled, but not painted. Instead of this, I'd suggest looking at the function signature and take the return type as the target class.